### PR TITLE
Option to format API Results as GeoJSON

### LINKED ISSUE #427: Add a `format` Param to the API
We should obviously default the response format to JSON, but we should also allow users to export data as GeoJSON via the API.

## Who is this for?
Mostly the CoCDO, but I can see this being useful to a lot of people.

## Technical Components
- [x] add a plug that handles `format`
- [x] plug defaults to `json`
- [x] add function to view code that reformats to GeoJSON where `conn.assigns[:format] = "geojson"`

## Acceptance Criteria
- [x] test "format defaults to json"
- [x] test "geojson is acceptable as format option"
- [x] test "format=geojson produces data as GeoJSON objects"
- [x] test "unacceptable format will 400"

Story: #426

### PR DESCRIPTION
**Pull request by vforgione**
The city data team actually uses this feature quite a bit with the current API, and I thought it's actually pretty slick.

Stuff in this PR:

- Fixed all of our wonky geom parsing to be correct `lon/lat` tuples
- Fixed bbox computing to use built in PostGIS functions to create a real bbox as a convex hull
- Added optional formatting param to the API

Closes #427

### CODE REVIEW

**Comment by vforgione** [comment]
If I weren't alright with it, I wouldn't have pushed it.

**Comment by HeyZoos** [comment]
Then shouldn't we change the name?

**Comment by brlodi** [comment]
BBoxes are specifically of the form `[min_lon, min_lat, max_lon, max_lat]`, so I agree we shouldn't be using that name (even internally) if we're not actually dealing with a true bbox.

**Comment by vforgione** [comment]
My rationale for changing it from an explicit rectangle to any polygon is that when users want the GeoJSON format for their responses for metadata, I would assume that they want to plot the _exact_ shape on a map.

I also would like to limit false positives in the explorer app -- naive rectangles could leave, and most often do, large blank spaces.

I'm down for leaving `bbox` and as a box and having a new `hull` field, but then I wouldn't know what the point of maintaining the box would be...

**Comment by HeyZoos** [comment]
I'm alright with ditching the box field.

**Comment by brlodi** [comment]
> GeoJSON format for their responses for metadata, I would assume that they want to plot the exact shape on a map.

I agree, but that's not a bbox so we shouldn't call it that (it doesn't look like you're sending it in the GeoJSON as such, but to avoid confusion we should also use a different name internally).

> I'm down for leaving bbox and as a box and having a new hull field, but then I wouldn't know what the point of maintaining the box would be...

BBox is useful for quick short-circuit intersection queries, so you can see if two complex shapes even possibly overlap before doing a more expensive operation. If we decide we don't care enough internally, it could still be useful to generate and provide on demand in the GeoJSON response.

**Comment by vforgione** [comment]
I kind of like the using the bbox and a simple, first pass filter and then refining using the hull. Overall it may be slightly overkill, but it may also cut out a lot of complexity involved with hull intersections.

We also need to add and idex for this (and `time_range` and `state`).

So, add a `hull` field to the `metas` table?

**Comment by vforgione** [comment]
Ok, I've added `hull` as its own field to both `Meta` and `AotMeta` and I've updated the general search strategy to start with bbox and refine with hull.

File: lib/plenario.ex [hunk 1]
```diff
@@ -4,46 +4,51 @@ defmodule Plenario do
...
-      end
-
-    Repo.all(query)
+  @spec search_data_sets(user_geom :: Geo.Polygon, time_range :: Postgrex.Range | nil) :: list(Meta)
+  def search_data_sets(user_geom, user_time_range \\ nil) do
+    search_bbox(user_geom, user_time_range)
+    |> Repo.all()
+    |> refine_to_hull(user_geom)
+    |> Repo.all()
+  end
```
Comment (HeyZoos) [inline lib/plenario.ex]: Not a necessary change, but I wonder if calculating intersections between complex geometries is expensive enough to justify prepping things with a bbox search. I like the idea of a zippy bbox search first, then a precise hull search second, but I wish I could find some _numbas_

Comment (brlodi) [reply lib/plenario.ex]: I don't know what the real world savings are for the dataset sizes we're working with, but comparing bboxes is O(1) while comparing hulls will be O(n*m) in the worst-case, where *n* and *m* are the number of vertices of the first and second polygon respectively.

**Pull request merged**